### PR TITLE
Implement NetworkInterfaceProperties

### LINKED ISSUE #22: Implement `io.edgehog.devicemanager.NetworkInterfaceProperties` Interface in Device Runtime
See also: https://github.com/edgehog-device-manager/edgehog-astarte-interfaces

### PR DESCRIPTION
**Pull request by FrancescoVaiani**
Add support to `io.edgehog.devicemanager.NetworkInterfaceProperties`

Close #22

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,6 +60,22 @@ state_file = "/var/lib/edgehog/state.json"
 download_directory = "/var/tmp/edgehog-updates/"
 ```
 
+## Dependencies
+
+In order to use the Edgehog Device Runtime, you must have a Linux system with the libudev library installed where it can be found by pkg-config. To install libudev on Debian-based Linux distributions, execute the following command:
```
Comment (harlem88) [inline README.md:65]: Move this part in the `Supported Operating System` section

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/telemetry/net_if_properties.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        let technology_type = match type_.to_string_lossy().trim() {
+            ARPHRD_ETHER => {
+                let uevent = device
+                    .attribute_value("uevent")
+                    .unwrap_or_default()
+                    .to_string_lossy()
+                    .into_owned();
+                if uevent.contains("DEVTYPE=wlan") {
+                    TechnologyType::WiFi
+                } else if uevent.contains("DEVTYPE=bridge") {
```
Comment (bettio) [inline src/telemetry/net_if_properties.rs:80]: I'm not sure if this is a good strategy, we might have other devices/interfaces that we may need to ignore, such as VPNs or docker interfaces.
Maybe a better filter would be based on the physical subsystem (such having a valid ID_BUS) or something like that. Anyway let's discuss this.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,17 @@ This is provided via the `rust-systemd` crate, a Rust interface to `libsystemd/l
 To build the `runtime` make sure you have `libsystemd-dev` installed on your system
 and the systemd feature enabled.
 
+To install `libudev` on Debian-based Linux distributions, execute the following command:
```
Comment (bettio) [inline README.md:22]: libudev development headers or something like that.

Comment (bettio) [reply README.md:22]: Also I would expand this, with something like:
"libudev is required for gathering information about network interface, in order to install libudev development headers ..." or something like this.

**Review by bettio — CHANGES_REQUESTED** [review]

File: src/telemetry/net_if_properties.rs (hunk 2 above)
Comment (harlem88) [reply src/telemetry/net_if_properties.rs:80]: Maybe we could distinguish between device virtual or not. I don't know if the libudev expose the virtual attribute. 
You can find all the virtual net device here:
`/sys/devices/virtual/net`

File: README.md [hunk 4]
```diff
@@ -19,6 +19,18 @@ This is provided via the `rust-systemd` crate, a Rust interface to `libsystemd/l
 To build the `runtime` make sure you have `libsystemd-dev` installed on your system
 and the systemd feature enabled.
 
+libudev is required for gathering information about network interface. In order ti install `libudev` development headers
```
Comment (harlem88) [inline README.md:22]: In order `ti` install -> In order `to`

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

File: src/telemetry/net_if_properties.rs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+#[derive(Debug)]
+struct NetworkInterfaceProperties {
+    interface: String,
+    mac_address: String,
+    technology_type: TechnologyType,
+}
+
+fn get_ethernet_wifi() -> Result<Vec<NetworkInterfaceProperties>, DeviceManagerError> {
```
Comment (rbino) [inline src/telemetry/net_if_properties.rs:51]: This actually returns also cellular, so I'd rather call it `get_supported_network_interfaces`

**Review by rbino — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -16,6 +16,27 @@ enables remote device management using
 ## Supported Operating System
 
 At the moment only Linux-based systems are supported.
+On Linux systems if `edgehog-device-runtime` is a `systemd` service, it can notify `systemd` of its status changes.
+This is provided via the `rust-systemd` crate, a Rust interface to `libsystemd/libelogind` APIs.
+
+To build the `runtime` make sure you have `libsystemd-dev` installed on your system
+and the systemd feature enabled.
```
Comment (harlem88) [inline README.md:23]: You reintroduced a section it has been moved to this page [os_requirements](https://github.com/edgehog-device-manager/edgehog-device-runtime/blob/main/doc/os_requirements.md).

File: README.md [hunk 7]
```diff
@@ -16,6 +16,27 @@ enables remote device management using
...
+
+For Red Hat distributions (e.g. Fedora):
+```shell
+sudo dnf install systemd-devel
+```
+
+In order to build the project with the systemd feature enabled execute the following command:
+```shell
+cargo build --features systemd
+```
```
Comment (harlem88) [inline README.md:39]: Same above

**Review by harlem88 — APPROVED** [review]

**Review by rbino — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -28,6 +28,7 @@ The following information are sent to remote Edgehog instance:
 - Runtime info and compiler version
 - OTA update using RAUC
 - `Edgehog Device Runtime` status changes via systemd.
+- Network interfaces info
```
Comment (bettio) [inline README.md:31]: `Network interface info`

File: doc/os_requirements.md [hunk 9]
```diff
@@ -13,6 +13,7 @@ Edgehog Device Runtime has a number of requirements in order to provide device m
 ### Dependencies
 * **Rust** >= 1.58
 * **libsystemd** (optional)
+* **libudev**: Gathering information about network interface.
```
Comment (bettio) [inline doc/os_requirements.md:16]: Needed for gathering ... `network interfaces`.

**Review by bettio — CHANGES_REQUESTED** [review]

**Review by bettio — APPROVED** [review]

**Pull request merged**